File: src/main/java/algo/slidingWindow/TakeKOfEachCharacterFromLeftAndRight_2516.java
package algo.slidingWindow;

public class TakeKOfEachCharacterFromLeftAndRight_2516 {
    public static int takeCharacters(String s, int k) {
        var count = new int[]{-1 * k, -1 * k, -1 * k};
        for (var i = 0; i < s.length(); i++) {
            count[s.charAt(i) - 'a']++;
        }
        if (count[0] < 0 || count[1] < 0 || count[2] < 0) {
            return -1;
        }
        if (count[0] == 0 && count[1] == 0 && count[2] == 0) {
            return s.length();
        }

        var maxLen = 0;
        int left = 0, right = 0;
        while (right < s.length()) {
            var c = s.charAt(right);
            count[c - 'a']--;
            while (left < right && (count[0] < 0 || count[1] < 0 || count[2] < 0)) {
                count[s.charAt(left) - 'a']++;
                left++;
            }
            maxLen = Math.max(maxLen, right - left + 1);
            right++;
        }
        return s.length() - maxLen;
    }

    public static void main(String[] args) {
        System.out.println(takeCharacters("aabaaaacaabc", 2));
        System.out.println(takeCharacters("a", 0));
        System.out.println(takeCharacters("abc", 1));
    }
}

File: src/main/kotlin/algo/slidingWindow/TakeKOfEachCharacterFromLeftAndRight_2516.kt
package algo.slidingWindow

class KtTakeKOfEachCharacterFromLeftAndRight_2516 {
    fun takeCharacters(s: String, k: Int): Int {
        val count = intArrayOf(-1 * k, -1 * k, -1 * k)  // upper bound of each char

        s.forEach { c -> count[c - 'a']++ }
        if (count[0] < 0 || count[1] < 0 || count[2] < 0) {
            return -1
        }
        if (count[0] == 0 && count[1] == 0 && count[2] == 0) {
            return s.length;
        }

        var left = 0; var right = 0
        var maxLen = 0
        while (right < s.length) {
            val c = s[right]
            count[c - 'a']--
            while (left < right && (count[0] < 0 || count[1] < 0 || count[2] < 0)) {
                // shrink the window from left
                count[s[left] - 'a']++
                left++
            }
            maxLen = maxLen.coerceAtLeast(right - left + 1)
            right++
        }
        return s.length - maxLen
    }
}

fun main() {
    val take = KtTakeKOfEachCharacterFromLeftAndRight_2516()
    println(take.takeCharacters("aabaaaacaabc", 2))
    println(take.takeCharacters("a", 0))
    println(take.takeCharacters("abc", 1))
}